File: LeetCodes/path-with-minimum-effort/9-16-2023, 9_08_39 AM/path-with-minimum-effort.cpp
// Link: https://leetcode.com/problems/path-with-minimum-effort

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: path-with-minimum-effort
 *  RUNTIME: 105 ms
 *  MEMORY: 20.3 MB
 *  DATE: 9-16-2023, 9:08:39 AM
 *
 */

// Solution:

class Solution {
public:
    int minimumEffortPath(vector<vector<int>>& heights) {
        int n = heights.size(), m = heights[0].size();

        priority_queue<pair<int,pair<int,int>>, vector<pair<int, pair<int, int>>>, greater<pair<int, pair<int, int>>>> pq;
        vector<vector<int>> minEffort(n, vector<int> (m, INT_MAX));
        int dirs[5] = {-1, 0, 1, 0, -1};

        pq.push({0, {0, 0}});
        minEffort[0][0] = 0;

        while(!pq.empty()) {
            int x = pq.top().second.first;
            int y = pq.top().second.second;
            int currEffort = pq.top().first;
            pq.pop();

            if(x == n-1 && y == m-1) {
                return currEffort;
            }

            for(int k=0; k<4; k++) {
                int row = x + dirs[k];
                int col = y + dirs[k+1];

                if(row >= 0 && col >= 0 && row < n && col < m) {
                    int newEffort = max(currEffort, abs(heights[x][y] - heights[row][col]));
                    if(newEffort < minEffort[row][col]) {
                        minEffort[row][col] = newEffort;
                        pq.push({newEffort, {row, col}});
                    }
                }
            }
        }
        return 0;
    }
};

File: LeetCodes/path-with-minimum-effort/5-15-2024, 5_51_46 PM/path-with-minimum-effort.java
// Link: https://leetcode.com/problems/path-with-minimum-effort

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: path-with-minimum-effort
 *  RUNTIME: 39 ms
 *  MEMORY: 45 MB
 *  DATE: 5-15-2024, 5:51:46 PM
 *
 */

// Solution:

class Solution {
    public int minimumEffortPath(int[][] heights) {
        int n = heights.length;
        int m = heights[0].length;

        Queue<int[]> pq = new PriorityQueue<>(new Comparator<int[]>() {
            public int compare(int[] a, int[] b) {
                if(a[0] == b[0]) return 0;
                return (a[0] > b[0]) ? 1 : -1;
            }
        });
        int[] dirs = new int[]{-1, 0, 1, 0, -1};
        int[][] efforts = new int[n][m];
        for(int i=0; i<n; i++) for(int j=0; j<m; j++) efforts[i][j] = Integer.MAX_VALUE;

        efforts[0][0] = 0;
        pq.add(make_pair(0, 0, 0));

        while(!pq.isEmpty()) {
            int[] node = pq.remove();
            int currEffort = node[0];
            int x = node[1];
            int y = node[2];

            if(x == n-1 && y == m-1)
                return currEffort;

            for(int k=0; k<4; k++) {
                int i = x + dirs[k];
                int j = y + dirs[k+1];

                if(i<0 || j<0 || i>=n || j>=m)
                    continue;
                int resEffort = Math.max(currEffort, Math.abs(heights[i][j] - heights[x][y]));
                if(resEffort < efforts[i][j]) {
                    efforts[i][j] = resEffort;
                   pq.add(make_pair(resEffort, i, j));
                }
            }
        }
        return -1;
    }

    public int[] make_pair(int a, int b, int c) {
        return new int[]{a, b, c};
    }
}
